### ISSUE TITLE
Sorting LSP suggestions with 'AI'

### ISSUE THREAD

**Issue by knyazer** [zed-industries/zed#16856]
### Check for existing issues

- [X] Completed

### Describe the feature

_Related to a bunch of features, but I believe this is significantly different from most of them_

Use the raw (not instruction tuned model) to do the ranking of the LSP suggestions.

Rank the suggestions based on the logits (likelihoods) of them under short context and some small very fast model. While doing it via an API is the simplest (e.g. request llama3.1 8b from some provider that gives logits), this is going to be pretty slow, since APIs are a lot slower than the LSPs: time to first token is around 200ms these days, and, if adding network latency to this it becomes unbearably slow. 

Thus running such a ranking model only makes sense if done locally. The problem becomes to construct and train a CPU-friendly small LLM that would have enough expressive ability to actually perform well. The training is simple: just distill the llama405b into it. The problem is actually spending time on this. And, even before that, we would need to check the feasibility, e.g. by writing up a plugin with llama8b and api, and checking that it actually is capable of suggesting things well.

WDYT?

### If applicable, add mockups / screenshots to help present your vision of the feature

![image](https://github.com/user-attachments/assets/11b77fa6-dee5-4cd6-9368-1ff76f7ddc12)

When I type ```o``` in this case, I want to see 'original_batch_size' as a suggestion, not whatever the LSP currently suggests (which is like 'or', 'of', 'os', 'os', 'os')

**Comment by knyazer** [comment]
By the way, if we could implement our own LLM to run locally this could be a great edge over the cursor editor: you could use LLMs while not sending your data anywhere, so using it as a corporate.

This actually sounds so exciting that I kinda want to work on microscopic LLMs myself. I remember a few years ago that would be a pain, but these days there is ```llama.cpp``` which is a nice runtime that supports a ton of stuff with minimal overhead, so it's not even that hard. The hardest part is to make these micro-llms actually good :)


I hope you guys collect enough data from zed to actually train something meaningful using it. I'm pretty sure cursor hoards just all the data one ever writes in it, since I don't see how you would otherwise build a "next cursor position prediction" model.

**Comment by notpeter** [comment]
Hi @knyazer, the idea of supporting a custom method for suggestions is interesting, I don't think tuned micro-llms are a great fit at this time.  At the moment there isn't great infrastructure to deploy such things (let alone tune them).  Additionally the speed is as you mentioned is a major concern. Given that suggestions often need to update with each additional keystroke that's 4-5 queries per second so to feel responsive it would have to always be sub-200ms, ideally sub 100ms (for all tokens, not just first token).  So this might be feasible in a few years, but not at time soon.

Interesting thought experiment, but I'm going to close as not-planned.
Thanks for taking the time to write this up.
